### PR TITLE
Refactors SSvote, makes votes into datums, also makes vote ui Typescript

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

- Makes vote into their own singleton datums.
- Refactors the voting subsystem to accommodate. 
- Refactors the vote UI from JS to TSX (probably badly). 
  - I'm not very good at typescript so review this part with scrutiny.

## Why It's Good For The Game

Having votes in modular, separate datums allows for much easier addition and removal of voting types, as well as de-hardcodes voting options and effects.

Instead of having multiple switches and if statements embedded in the vote SS for each vote type, new votes can be implemented on a per-datum basis. This also allows for easier and stronger configuration of votes, such as being able to config-disable a vote in its entirety (even from admins) (not implemented in this pr - but allows it). 

## Changelog

:cl: Melbert
refactor: Refactored the voting subsystem. Votes are now each their own datums. 
refactor: The voting UI now uses Typescript.
/:cl:

### CODE REVIEW

**Review by stylemistake — APPROVED** [review]

**Comment by stylemistake** [comment]
Good I think.

**Comment by tralezab** [comment]
i also took a look and liked it ;)

File: code/datums/votes/map_vote.dm [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	for(var/map in choices)
+		var/datum/map_config/possible_config = config.maplist[map]
+		if(possible_config.config_min_users > 0 && GLOB.clients.len < possible_config.config_min_users)
+			choices -= map
+
+		else if(possible_config.config_max_users > 0 && GLOB.clients.len > possible_config.config_max_users)
+			choices -= map
+
+/datum/vote/map_vote/toggle_votable(mob/toggler)
+	if(!check_rights_for(toggler?.client, R_ADMIN))
```
Comment (tralezab) [inline code/datums/votes/map_vote.dm:32]: why would toggler not exist? are you sure this isn't suppressing a runtime?

File: code/datums/votes/map_vote.dm [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+			// Non-voters will have their preferred map voted for automatically.
+			var/their_preferred_map = non_voter_client?.prefs.read_preference(/datum/preference/choiced/preferred_map)
+			var/default_map = global.config.defaultmap.map_name
+			// If the non-voter's preferred map is null for some reason, we just use the default map.
+			var/voting_for = their_preferred_map || default_map
+
+			if(voting_for in choices)
+				choices[voting_for] += 1
+
+	return ..()
```
Comment (tralezab) [inline code/datums/votes/map_vote.dm:72]: invert this condition to early return

File: code/datums/votes/map_vote.dm [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+			if(by_who)
+				to_chat(by_who, span_warning("Map voting is disabled."))
+			return FALSE
+
+		if(SSmapping.map_voted)
+			if(by_who)
+				to_chat(by_who, span_warning("The next map has already been selected."))
+			return FALSE
+
+	return TRUE
```
Comment (tralezab) [inline code/datums/votes/map_vote.dm:57]: invert this condition to early return

**Review by tralezab — CHANGES_REQUESTED** [review]
very little things stuck out to me, this is so well written lol

File: code/datums/votes/map_vote.dm (hunk 1 above)
Comment (MrMelbert) [reply code/datums/votes/map_vote.dm:32]: I intended the "`toggler?`" check as an anti-spoof check for someone calling it when they shouldn't - but, I suppose by extension it's also silencing a runtime

File: tgui/packages/tgui/interfaces/VotePanel.tsx [hunk 4]
```diff
@@ -0,0 +1,201 @@
+import { BooleanLike } from 'common/react';
+import { Box, Icon, Stack, Button, Section, NoticeBox, LabeledList, Collapsible } from '../components';
+import { Window } from '../layouts';
+import { useBackend } from '../backend';
+
+type Vote = {
+  name: string;
+  canBeInitiated: BooleanLike;
+  config: number; // -1 = no config, 0 = config + disabled, 1 = config + enabled
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/VotePanel.tsx:9]: Make this an enum type instead, like:

```ts
enum VoteConfig {
    None = -1,
    Disabled = 0,
    Enabled = 1,
}
```

Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/VotePanel.tsx:9]: You should then change every reference of 0, 1, etc to `VoteConfig.Enabled`, etc.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/VotePanel.tsx (hunk 4 above)
Comment (tralezab) [reply tgui/packages/tgui/interfaces/VotePanel.tsx]: ENUMSSSSS YESSS

**Review by tralezab — APPROVED** [review]
cool B)

**Pull request merged**